third_party/amd/lib/TritonAMDGPUTransforms/CanonicalizePointers.cpp

```diff
@@ -442,16 +268,118 @@ PointerCanonicalizer::decomposeOffsetFromExpr(Location loc, Value expr,
   return offsets;
 }
 
-Value PointerCanonicalizer::createTensorPointer(FatPtr fatPtr, Location loc) {
-  Value basePtr = fatPtr.basePtr;
-  Value offset = fatPtr.offset;
+static const std::string kPtrCanonPrefix = "__amdpointercanonicalize.";
+static const std::string kSCFThenRewrittenAttr =
+    kPtrCanonPrefix + "scf-then-rewritten__";
+static const std::string kSCFElseRewrittenAttr =
+    kPtrCanonPrefix + "scf-else-rewritten__";
+static const std::string kSCFIfOpYieldFatPtrOffsets =
+    kPtrCanonPrefix + "scf-if-yield-fatptr-offsets__";
+
+struct FatPointers {
+  struct FatPtrAttrs {
+    FatPtrAttrs(const FatPtrAttrs &other) = default;
+    FatPtrAttrs &operator=(const FatPtrAttrs &other) = default;
+    // for map default insert
+    FatPtrAttrs() = default;
+    bool canNarrow = false;
+    llvm::SmallDenseMap<StringRef, Attribute> attributes;
+    friend bool operator==(const FatPtrAttrs &lhs, const FatPtrAttrs &rhs) {
+      return lhs.canNarrow == rhs.canNarrow && lhs.attributes == rhs.attributes;
+    }
+    friend bool operator!=(const FatPtrAttrs &lhs, const FatPtrAttrs &rhs) {
+      return !(lhs == rhs);
+    }
```

Comment:
Style: Can we reorganize it a bit so that we have constructors first, and then operators, and then data fields. Use some empty lines to separate them.

Similarly for the outer class. A bit better for readability.

Reply:
As we discussed during 1-1, I might actually remove this thing entirely (depending on whether `int-range-narrow` works for us). But yea sure I can re-org here.